text/0079-exception-groups.md

```diff
@@ -0,0 +1,348 @@
+- There can be a stack trace on each of the exceptions within an exception group, as well as on the group itself.
+- Just like any other exception, each exception within an exception group can have a chain of inner exceptions.
+- An inner exception can *also* be an exception group.
+
+Thus, other valid examples that could occur in an application include the following:
+
+- Group B
+  - Exception 1
+    - Exception 2
+      - Exception 3
+
+- Group C
+  - Exception 1
+    - Exception 1a
+      - Exception 1b
+  - Exception 2
+    - Exception 2a
+      - Exception 2b
+
+- Group D
+  - Exception 1
+    - Exception 1a
+      - Exception 1b
+  - Group E
+    - Exception 2
+      - Exception 2a
+        - Exception 2b
+    - Exception 3
+      - Exception 3a
+        - Exception 3b
+
+## Exception Handling in Sentry
+
+SDKs send exceptions to Sentry using the [Exception Interface](https://develop.sentry.dev/sdk/event-payloads/exception/)
+on an event. One or more exceptions can be sent in the `values` array of the interface.  Multiple values represent
+a chain of exceptions in a causal relationship, sorted from oldest to newest.  For example:
+
+```json
+{
+  "exception": {
+    "values": [
+      {"type": "TypeError", "value": "Invalid Type!"},
+      {"type": "ValueError", "value": "Invalid Value!"},
+      {"type": "RuntimeError", "value": "Something went wrong!"}
+    ]
+  }
+}
+```
+
+In the above example, an issue is created in Sentry for the `RuntimeError` exception.
+
+```
+Issue Title: "RuntimeError: Something went wrong!"
+
++==============================
++ RuntimeError: Something went wrong!
++------------------------------
++ (stack trace)
++==============================
+
++==============================
++ ValueError: Invalid Value!
++------------------------------
++ (stack trace)
++==============================
+
++==============================
++ TypeError: Invalid Type!
++------------------------------
++ (stack trace)
++==============================
+```
+
+This design is linear, and thus cannot support exception groups having more than one exception.
+But even in that case, the exception presented in Sentry would not be titled or grouped correctly.
+
+Consider:
+
+```json
+{
+  "exception": {
+    "values": [
+      {"type": "TypeError", "value": "Invalid Type!"},
+      {"type": "ValueError", "value": "Invalid Value!"},
+      {"type": "RuntimeError", "value": "Something went wrong!"},
+      {"type": "exception_group", "value": "Exception Group (1 sub-exception)"}
+    ]
+  }
+}
+```
+
+Sending this to Sentry would result in an issue titled `"exception_group: Exception Group (1 sub-exception)"`,
+which is unactionable.
+
+# Supporting Data
+
+- [Python customer feedback](https://github.com/getsentry/sentry/issues/37716)
+- [.NET customer feedback](https://github.com/getsentry/sentry-dotnet/issues/270)
+
+# Options Considered
+
+## Option 1
+
+Do nothing.
```

Comment:
This is an outright question: Would it make sense to add options 1b and 1c that contain:
- SDK reports each exception in a group as separate error event; no change to protocol.
- Advice to developers (customers) that they should manually inject a `capture_exception` at the places where the groups are caught and capture them separately. This is particularly to avoid cases where a group is nested.

Additionally, can it happen that _one_ of the exceptions in the group can be viewed as "causal" for a chain whereas other exceptions are just additional baggage? In this case, we would definitely need SDK-side mapping to determine which exception should be separated out.

Reply:
Yeah, that doesn't work out very well in practice.  I'll add some information and reasons why to the RFC.

Reply:
On the second question, if only one is the cause then they're only be one exception in the group to begin with.  However, usually an exception group is raised if *any* of a list of tasks fails. Thus it can be possible that one of the exceptions is the true cause, and the others are noise created by cancelling a running task (if they're aborted). There's really no way to tell - so we have to treat all of them as potential causes.

Reply:
Actually, I forgot one thing.  At least for .NET an `AggregateException` will never contain an `OperationCanceledException`.  So if say two tasks are being waited upon, and one of them throws some exception and the other is cancelled, only the one exception will be in the group.  And if _both_ tasks are cancelled, then an `OperationCanceledException` will be thrown _instead of_ an `AggregateException`.

So ultimately, this aligns with the idea of all exceptions in an exception group _together_ are the cause for the exception group.  There isn't a case where some are additional baggage.  Not really.

Though also, sometimes you might not care about the others.  Even if they're meaningful, they might be all be similar.  We could have an option to discard all but one of them.  I'll add a few more sub-options to the list.  Thanks.

Reply:
The RFC has been changed considerably. The revised proposal describes the "SDK Splitting" approach under the "Other Options Considered" section.  Thanks.